### PR TITLE
sys/bitfield: add support for bit-wise bitfield operations

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This adds support for bit-wise (AND, OR, XOR, NOT) operations to bitfields.

### Testing procedure

The `tests-bitfield` unit test has been extended with the new operations.

### CODE REVIEW

File: sys/include/bitfield.h [hunk 1]
```diff
@@ -89,6 +89,69 @@ static inline bool bf_isset(uint8_t field[], size_t idx)
...
 
+/**
+ * @brief  Perform a bitwise OR operation on two bitfields
+ *         `out = a | b`
+ *
+ * @param[out]    out   The resulting bitfield
+ * @param[in]     a     The first bitfield
+ * @param[in]     b     The second bitfield
+ * @param[in]     len   The size of the bitfield in bytes
+ */
```
Comment (OlegHahm) [inline sys/include/bitfield.h:100]: I think we should add a `@pre` stating that the size for a, b, and c must be greater than `len`.

Also, I would rephrase to something like
```
  * @param[in]     len   The number of bytes to which the operation is applied.
```

Comment (benpicco) [reply sys/include/bitfield.h:100]: > the size for a, b, and c must be greater than `len`.

Shouldn't it be *at least* `len`? 
I assumed one would typically run the operation on the whole bitfield, not arbitrary byte-sized chunks.

Comment (OlegHahm) [reply sys/include/bitfield.h:100]: Sure, it should just be documented exactly to make sure that it is the responsibility of the caller to check these preconditions.

Comment (benpicco) [reply sys/include/bitfield.h:102]: I feel like this is stating the obvious, but [there](https://github.com/RIOT-OS/RIOT/compare/5c0cb8bed44e88f853fcfdd7919a7122020a7a7a..c55cd51e8ef52acc74523dae20bda4b293a39d80) you go

**Review by kfessel — COMMENTED** [review]
Exiting functions/macros use `size` as the number of bits. 

There already is an operation `int bf_get_unset(uint8_t field[], int size);`  which has a size parameter, they should all use the same type - seems `size_t`  would be correct.

`bitfield.h` also seems to miss some kind of initialization illustrated by the example l202 and l203

**Comment by kaspar030** [comment]
> There already is an operation `int bf_get_unset(uint8_t field[], int size);` which has a size parameter, they should all use the same type - seems `size_t` would be correct.

I agree on this. Also, using bits vs. bytes doesn't fail as horrible as if someone supplies the number of bits as bytes arbument. (does this sentence make sense?)

**Comment by kaspar030** [comment]
I'd like to see a line or two on the unused bits of a bitfield and how they are affected by the ops. I mean, a "bitfield" of 10 bits uses 2 bytes of storage. the unused 6 bits will get modified by the current and/or/xor implementations. That is fine IMO, but should be documented.

File: sys/bitfield/bitfield.c [hunk 2]
```diff
@@ -21,11 +21,11 @@
 #include "bitfield.h"
 #include "irq.h"
 
-int bf_get_unset(uint8_t field[], int size)
+int bf_get_unset(uint8_t field[], size_t size)
 {
     int result = -1;
     int nbytes = (size + 7) / 8;
-    int i = 0;
+    size_t i = 0;
```
Comment (kfessel) [inline sys/bitfield/bitfield.c:28]: l43: some static test likes to see a space between `return` and `(`

File: sys/include/bitfield.h [hunk 3]
```diff
@@ -89,18 +89,105 @@ static inline bool bf_isset(uint8_t field[], size_t idx)
...
 }
 
+/**
+ * @brief  Perform a bitwise OR operation on two bitfields
+ *         `out = a | b`
+ *
+ * @pre   The size of @p a, @p b and @p out must be at least @p len bits
+ *
+ * @note  This operation will also affect unused bits the bytes that make up
+ *        the bitfield.
```
Comment (kfessel) [inline sys/include/bitfield.h:99]: these notes are missing some preposition between `bits` and `the bytes`
```suggestion
 * @note  This operation will also affect unused bits of the bytes that make up
 *        the bitfield.
```

**Review by kfessel — APPROVED** [review]
murdock test are sufficient in this case, and murdock is green.
